### PR TITLE
Remove RwLockReadGuard usages

### PR DESCRIPTION
**Pull request by will118**
## Bug

Locks in wasm are assumed to be single threaded without concurrency: https://github.com/rust-lang/rust/blob/6f8c0557e0b73c73a8a7163a15f4a5a3feca7d5c/library/std/src/sys/unsupported/locks/rwlock.rs#L4

However, with Cloudflare Workers the situation is not so straightforward.

A single isolate can process multiple requests concurrently (not in parallel), so holding ReadGuards across await points is not OK.

There are some official docs that refer to the behaviour, but I must admit even having read those I wasn't convinced until I reproed it myself: https://developers.cloudflare.com/workers/learning/how-workers-works/#distributed-execution

The errors only happen under load, likely because the Workers runtime will readily spawn additional isolates.

The errors seen are:

    Error: The script will never generate a response.

    and

    Memory access out of bounds

I am reasonably confident this fixes the first, and somewhat confident it fixes the latter.

## Fix

The fix currently just clones the value instead of passing out the `ReadGuard`.

There are some accesses to the `key()` which (along with minimizing changes) is why I changed the API as little as possible.

### CODE REVIEW

**Review by xtuc — APPROVED** [review]

**Review by cjpatton — APPROVED** [review]
Wow, that is really tricky! Good work. :duck: - but I think the documentation can be improved slightly, see inline comment.

File: daphne_worker/src/config.rs [hunk 1]
```diff
@@ -558,26 +558,28 @@ impl<'srv> DaphneWorker<'srv> {
         Ok(None)
     }
 
+    /// In memory cache on-top of KV to avoid hitting KV API limits.
+    // NB: We shouldn't return guards from this function as they could be held across await points.
```
[the remark this answers is not in the document]
Comment (bhalleycf) [reply daphne_worker/src/config.rs:562]: "nota bene" I think!

[the remark this answers is not in the document]
Comment (will118) [reply daphne_worker/src/config.rs]: yep, I've changed it to "NOTE" now anyway.

File: daphne_worker/src/config.rs [hunk 2]
```diff
@@ -558,26 +558,28 @@ impl<'srv> DaphneWorker<'srv> {
...
+    //     Refer to commit msg for more details.
```
[the remark this answers is not in the document]
Comment (will118) [reply daphne_worker/src/config.rs]: suits me :)

**Pull request merged**